[libc++] Speed up set_intersection() by fast-forwarding over ranges of non-matching elements with one-sided binary search.

ichaer opened the pull request:
One-sided binary search, aka meta binary search, has been in the public domain for decades, and has the general advantage of being constant time in the best case, with the downside of executing at most 2\*log(N) comparisons vs classic binary search's exact log(N). There are two scenarios in which it really shines: the first one is when operating over non-random-access iterators, because the classic algorithm requires knowing the container's size upfront, which adds N iterator increments to the complexity. The second one is when traversing the container in order, trying to fast-forward to the next value: in that case the classic algorithm requires at least O(N\*log(N)) comparisons and, for non-random-access iterators, O(N^2) iterator increments, whereas the one-sided version will yield O(N) operations on both counts, with a best-case of O(log(N)) comparisons which is very common in practice.

Benchmark results for libcxx/benchmarks/set_intersection.libcxx.out on the following system:
Run on (16 X 4800.01 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x8)
  L1 Instruction 32 KiB (x8)
  L2 Unified 1280 KiB (x8)
  L3 Unified 24576 KiB (x1)

Comparison between main branch at b926f75e89 and this branch at be6c5c848d, with a fixed seed of 0 hardcoded in `getRandomEngine()` for consistent and reproducible results -- the full output has 1.3k lines, so I'm only pasting here the lines with best and worst CPU time differences, plus the summary:
```
Benchmark                                                Time      CPU     Time Old     Time New   CPU Old   CPU New
--------------------------------------------------------------------------------------------------------------------
SetIntersection_None_Vector_uint32_262144_262144      -0.0210  -0.7069  13352003320  13071795797   1204337    352960
SetIntersection_Interlaced_Set_uint64_262144_262144   -0.0207  +0.0459  13404448283  13127521495  53605744  56068263
OVERALL_GEOMEAN                                       -0.0222  -0.1220            4            4         0         0
```

ichaer committed:
[libc++][test] Add lower_bound complexity validation tests prior to introducing one-sided binary search for non-random iterators.

ichaer committed:
[libc++] Introduce one-sided binary search for lower_bound on non-random iterators.

One-sided binary search, aka meta binary search, has been in the public domain for decades, and has the general
advantage of being Ω(1) rather than the classic algorithm's Ω(log(n)), with the downside of executing at most
2*log(n) comparisons vs the classic algorithm's exact log(n). There are two scenarios in which it really shines:
the first one is when operating over non-random iterators, because the classic algorithm requires knowing the
container's size upfront, which adds Ω(n) iterator increments to the complexity. The second one is when you're
traversing the container in order, trying to fast-forward to the next value: in that case, the classic algorithm
would yield Ω(n*log(n)) comparisons and, for non-random iterators, Ω(n^2) iterator increments, whereas the one-sided
version will yield O(n) operations on both counts, with a Ω(log(n)) bound on the number of comparisons.

ichaer committed:
[libc++][test] Add set_intersection complexity validation tests prior to introducing use of one-sided binary search to fast-forward over ranges of elements.

ichaer committed:
[libc++] Introduce use of __lower_bound_onesided to improve average complexity of set_intersection.

mordante commented:
Thanks for the patch. If I understand it right this might violate the complexity requirements in http://eel.is/c++draft/lower.bound#4. Is that correct?

ichaer commented:
Hi Mark, thanks for your reply!

Strictly speaking, yes, that is true: the number of comparisons may violate the requirement of there being at most log(last - first) + O(1) projections and comparisons. However, that is only being done for non-random access iterators, where we are also executing 2*(last-first) iterator mutations, which are not part of the explicit algorithm guarantees, and are the reason why std::set and std::map have their own lower_bound -- std::lower_bound is really hard to use on non-random iterators because of this additional cost.

Would this be a blocker for accepting the contribution?

philnik777 commented:
> Hi Mark, thanks for your reply!
> 
> Strictly speaking, yes, that is true: the number of comparisons may violate the requirement of there being at most log(last - first) + O(1) projections and comparisons. However, that is only being done for non-random access iterators, where we are also executing 2*(last-first) iterator mutations, which are not part of the explicit algorithm guarantees, and are the reason why std::set and std::map have their own lower_bound -- std::lower_bound is really hard to use on non-random iterators because of this additional cost.
> 
> Would this be a blocker for accepting the contribution?

Yes. We have to stay standards-conforming - libc++ is an implementation of the C++ standard after all, so even if the trade-off seems like a good idea we can't make the call in this case. If this improves the performance for trivial calls like an integer and `std::less` or friends as the comparator, that should be allowable, since that's not user-visible. Anything as generic as you currently have won't be possible.

ichaer commented:
OK, that's fair. How about I just stop exposing one-sided lower bound through std::lower_bound, and just use it for std::set_intersection, where all complexity guarantees are preserved, would that be OK?

philnik777 commented:
> OK, that's fair. How about I just stop exposing one-sided lower bound through std::lower_bound, and just use it for std::set_intersection, where all complexity guarantees are preserved, would that be OK?

If all the complexity guarantees are met, that would be OK. We always like to see some benchmarks to have some data that the optimization actually improves performance.

mordante commented:
> SGTM! I'll amend my submission, thanks!

Great. 

As @philnik777 said we need to be careful about observable behaviour. I love to see performance improvements :-)

ichaer commented:
Cool, thanks :). Just for context, this code is actually being used in production out there. I work for Splunk, and we've implemented this outside of the standard library to speed up a component that does a number of set intersections, overlaying containers on top of each other (it's more complex, but that's the relevant part in a nutshell). The huge majority of the sets involved contain a tiny subset of what the target contains, so this makes a huge difference, and we thought we should contribute upstream.

Obviously I won't ask you to take me on my word :). Code is truth, I'll spend a bit of time understanding existing benchmarks in the LLVM repo to include something relevant in the pull request.

philnik777 commented:
I don't think you're wrong that this improves performance, but being able to point to hard data is always great. With algorithms it's almost always really easy to write micro benchmarks, so there we really like to have them.

ichaer commented:
Sounds good to me! I'll be out on holiday for a couple of weeks and should be able to update this PR in early January. Thanks for your feedback so far, I appreciate it :).

ichaer committed:
Fix `constexpr` annotations.

ichaer committed:
Remove std::exchange dependency from std::set_intersection so it works before C++14

ichaer committed:
Review feedback: don't use one-sided lower bound in lower_bound() itself since that violates the complexity guarantees from the standard.

ichaer committed:
Create new benchmark for set_intersection().

ichaer committed:
Formatting fixups.

ichaer committed:
General improvements to benchmark, including simplifying and slimming it down for faster runs, and including comparison counter.

ichaer committed:
Huh, I wonder how I got `git clang-format` to miss those changes =/

ichaer commented on libcxx/include/__algorithm/lower_bound.h:
```diff
@@ -46,6 +48,43 @@ _Iter __lower_bound(_Iter __first, _Sent __last, const _Type& __value, _Comp& __
+// version will yield O(n) operations on both counts, with a Ω(log(n)) bound on the number of comparisons.
+template <class _AlgPolicy, class _Iter, class _Sent, class _Type, class _Proj, class _Comp>
+_LIBCPP_NODISCARD_EXT _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 _Iter
+__lower_bound_onesided(_Iter __first, _Sent __last, const _Type& __value, _Comp& __comp, _Proj& __proj) {
```
I left `__lower_bound_onesided()` here even after removing all references to it from this file because it looked like it would be happier (and easier to find and maintain) with its family. I don't feel strongly about it, though.

ichaer commented on libcxx/test/std/algorithms/alg.sorting/alg.binary.search/lower.bound/lower_bound.pass.cpp:
I've preserved the test improvements I made here and to `lower_bound_comp.pass.cpp` because it seemed a shame to waste them. Please let me know if you disagree or would prefer for those changes to be split into a different pull request.

philnik777 replied:
Let's move them to a different PR.

ichaer replied:
SGTM, removed from this PR, will propose a new one with those test changes once we either merge or reject this one, because they depended on `stride_counting_iterator` changes.

ichaer committed:
Oops, bad mistake while porting into libc++! `__lower_bound_onesided()` must start with `__step==0`, otherwise we can't match the complexity of linear search when continually matching (like a std::set_intersection() of matching containers will).

ichaer committed:
Oops, bad tracking of displacement on `stride_counting_iterator`

ichaer committed:
Add more counters to the set_intersection benchmark, guard them behind an environment variable so we can choose to either measure time more accurately or obtain more information.

This led me down an interesting road of validating benchmark results and finding a significant discrepancy in timings between when I run all test cases at once or `--benchmark-filter` them individually.

ichaer commented:
@philnik777, @mordante, perhaps you'll have stumbled on this in the past: I was getting surprising results while running benchmarks, for instance I would consistently measure improvements across the board for `std::vector<uint32_t>`, but worse performance throughout for `std::vector<uint64_t>`. So I went run `perf stat` filtering for those specific test cases to check if it really was cache locality (my best guess), and results were completely different! Performance of my version was shown as better in the huge majority of the cases. What I then did was collect test names and re-run the process for each of them separately, and after that results started making sense to me.

Have you seen this happening before? I found one bug report for Google Benchmark which might potentially be related, https://github.com/google/benchmark/issues/461, but it's unresolved and very unclear. Another one is https://github.com/google/benchmark/issues/1469, for which `--benchmark_enable_random_interleaving` was created, but to use that I'd have to move away from internal batching, which is the normal way of doing it in the libc++ benchmarks I looked at, and which seems to be overall better -- I was doing some debugging and found the benchmark going through init steps again when using `--benchmark_repetitions`.

In any case, time-wise, benchmark results are largely positive, with worse results predominantly when (1) the cost of iterating is greater than the cost of comparison (`std::set` and cheap-to-compare types), and (2) when both input sets are identical, in which case all the additional machinery is overhead.

In terms of operation counts, however... I think the results speak for themselves =D. I'm obviously biased, but, to me, this is a thing of beauty. If the priority is scalability, this looks like an easy choice to me.

I'm attaching results I've tabulated in the way I described, filtering such that each is executed in an independent process. The results labeled as "original" are from 142e567cf008, "onesided" is from the current tip of this branch. I've also executed benchmarks twice for each version, to collect timings without the instrumentation required for the additional operation counters.

[comparison.ods](https://github.com/llvm/llvm-project/files/14041313/comparison.ods)
[comparison.pdf](https://github.com/llvm/llvm-project/files/14041314/comparison.pdf)

ichaer committed:
Merge branch 'main' into onesided_lower_bound

* main: (5908 commits)
  [readtapi] Cleanup printing command line options (#75106)
  [flang] Fix compilation error due to variable no being used (#75210)
  [C API] Add getters and setters for fast-math flags on relevant instructions (#75123)
  [libc++][CI] Tests the no RTTI configuration. (#65518)
  [RemoveDIs] Fold variable into assert, it's only used once. NFC
  [RemoveDI] Handle DPValues in SROA (#74089)
  [AArch64][GlobalISel] Test Pre-Commit for Look into array's element
  [mlir][tensor] Fix bug in `tensor.extract(tensor.from_elements)` folder (#75109)
  [analyzer] Move alpha checker EnumCastOutOfRange to optin (#67157)
  [RemoveDIs] Handle DPValues in replaceDbgDeclare (#73507)
  [SHT_LLVM_BB_ADDR_MAP] Implements PGOAnalysisMap in Object and ObjectYAML with tests.
  [X86][GlobalISel] Add instruction selection for G_SELECT (#70753)
  [AMDGPU] Remove unused function splitScalar64BitAddSub
  [LLVM][DWARF] Add compilation directory and dwo name to TU in dwo section (#74909)
  [clang][Interp] Implement __builtin_ffs (#72988)
  [RemoveDIs] Update ConvertDebugDeclareToDebugValue after #72276 (#73508)
  [libc][NFC] Reuse `FloatProperties` constant instead of creating new ones (#75187)
  [RemoveDIs] Fix removeRedundantDdbgInstrs utils for dbg.declares (#74102)
  Reapply "[RemoveDIs][NFC] Find DPValues using findDbgDeclares  (#73500)"
  [GitHub] Remove author association print from new-prs workflow
  ...

ichaer committed:
Merge branch 'main' into onesided_lower_bound

ichaer committed:
Revert "Oops, bad tracking of displacement on `stride_counting_iterator`"

This reverts commit 995d04b872c8552633c36e38d382897e8329d1e2.

ichaer committed:
* Fix C++03 compatibility issues.
* Fix tests I had broken.
* More tweaks and better comments.

ichaer committed:
Remove non-ascii characters, CI doesn't like them.

ichaer committed:
Merge remote-tracking branch 'llvm/main' into onesided_lower_bound

* llvm/main: (500 commits)
  [docs] Add beginner-focused office hours (#80308)
  [mlir][sparse] external entry method wrapper for sparse tensors (#80326)
  [StackSlotColoring] Ignore non-spill objects in RemoveDeadStores. (#80242)
  [libc][stdbit] fix return types (#80337)
  Revert "[RISCV] Refine cost on Min/Max reduction" (#80340)
  [TTI]Add support for strided loads/stores.
  [analyzer][HTMLRewriter] Cache partial rewrite results. (#80220)
  [flang][openacc][openmp] Use #0 from hlfir.declare value when generating bound ops (#80317)
  [AArch64][PAC] Expand blend(reg, imm) operation in aarch64-pauth pass (#74729)
  [SHT_LLVM_BB_ADDR_MAP][llvm-readobj] Implements llvm-readobj handling for PGOAnalysisMap. (#79520)
  [libc] add bazel support for most of unistd (#80078)
  [clang-tidy] Remove enforcement of rule C.48 from cppcoreguidelines-prefer-member-init (#80330)
  [OpenMP] Fix typo (NFC) (#80332)
  [BOLT] Enable re-writing of Linux kernel binary (#80228)
  [BOLT] Adjust section sizes based on file offsets (#80226)
  [libc] fix stdbit include test when not all entrypoints are available (#80323)
  [RISCV][GISel] RegBank select and instruction select for vector G_ADD, G_SUB (#74114)
  [RISCV] Add srmcfg CSR from Ssqosid extension. (#79914)
  [mlir][sparse] add sparsification options to pretty print and debug s… (#80205)
  [RISCV][MC] MC layer support for the experimental zalasr extension (#79911)
  ...

ichaer committed:
Oops, missed an #include

ichaer commented:
Sorry for all the noise, I'll work through the CI failures in a private PR... I thought I had a good handle on the tests being executed, but clearly I was mistaken :sweat_smile:

ichaer committed:
Merge remote-tracking branch 'llvm/main' into onesided_lower_bound

* llvm/main: (328 commits)
  [Flang][OpenMP] Attempt to make map-types-and-sizes.f90 test more agnostic to other architectures
  [Transforms] Add more cos combinations to SimplifyLibCalls and InstCombine (#79699)
  [workflows] Close issues used for backports once the PR has been created (#80394)
  [RISCV] Add support for RISC-V Pointer Masking (#79929)
  [lldb] Cleanup regex in libcxx formatters (NFC) (#80618)
  [lldb] Remove unused private TypeCategoryMap methods (NFC) (#80602)
  [mlir][sparse] refine sparse assembler strategy (#80521)
  [NFC] Fix typo (#80703)
  Fix broken ARM processor features test (#80717)
  [ValueTracking][NFC] Pass `SimplifyQuery` to `computeKnownFPClass` family (#80657)
  [x86_64][windows][swift] do not use Swift async extended frame for wi… (#80468)
  [X86] addConstantComments - add FP16 MOVSH asm comments support
  [X86] Regenerate some vector constant comments missed in recent patches to improve mask predicate handling in addConstantComments
  [clang][AMDGPU][CUDA] Handle __builtin_printf for device printf (#68515)
  Add some clarification to email check message
  [GitHub][Workflows] Prevent multiple private email comments (temporarily) (#80648)
  [workflows] Use /mnt as the build directory on Linux (#80583)
  [Flang][OpenMP] Initial mapping of Fortran pointers and allocatables for target devices (#71766)
  [AMDGPU] GlobalISel for f8 conversions (#80503)
  [AMDGPU] Fixed byte_sel of v_cvt_f32_bf8/v_cvt_f32_fp8 (#80502)
  ...

ichaer committed:
set_intersection.h: remove `static constexpr`, it breaks constexprness of the method.

ichaer committed:
Fix constexpr shenanigans with gcc and `stride_counting_iterator`

ichaer committed:
Restrict number of constexpr steps so `ranges_set_intersection.pass.cpp` is ok on gcc.

ichaer committed:
Fix constexpr annotation and make internal methods private in _IterOps

ichaer committed:
Allow for assertions in comparison count when in hardened mode for complexity validation.

ichaer commented:
@philnik777, I'm convinced that the one test failure I got is unrelated, and, from the looks of it, a likely flaky test (`std/thread/thread.semaphore/timed.pass.cpp`, in the `mingw-dll-i686` job). Could you please take a look at the changes now?

ichaer committed:
Revert lower_bound.pass.cpp changes, will move into a new PR.

ichaer committed:
Oops, forgot to revert this one too.

ichaer commented:
@philnik777, I'd love to hear your thoughts about the change :)

ichaer commented:
@philnik777, any more feedback?

ichaer commented:
@philnik777, what do you think? Any more feedback?

philnik777 reviewed (commented):
Sorry for being so slow on this. I don't really have the time currently to review this, since it's quite complex. @ldionne could you take a look?

ichaer commented on libcxx/include/__algorithm/iterator_operations.h:
```diff
@@ -85,6 +86,56 @@ struct _IterOps<_ClassicAlgPolicy> {
   }
 
+  // advance with sentinel, a la std::ranges::advance
+  // it's unclear whether _Iter has a difference_type and whether that's signed, so we play it safe:
```
I was trying to preserve compatibility with the previous implementation of `std::set_intersection()`, which didn't make use of `difference_type`. I completely agree with you that the standard requires `LegacyInputIterator`, which mandates the presence of `difference_type`, but it felt like a small price to pay to preserve the flexibility of the previous implementation...

I'm happy to tighten the requirements if you're sure.

ichaer replied:
I think maybe I understand where you're coming from, are you concerned about this not being a faithful version of `std::advance()`, and the similarity confusing others into making mistakes? When I wrote it like that, I was thinking that having a less strict fallback for the function when operating in `_ClassicAlgPolicy` would be a net positive, because it allows us to preserve pre-existing behaviour when using it, but I think I understand your point of view. Personally, I would probably leave it as it is (it's targeted at libc++ implementors and the whole thing will be inlined anyway), but if you disagree, I have two alternatives I'd like to suggest:
1. Rename this to something like `advance_with_custom_difference_type()` so that the intent is explicit, and add a comment explaining that its goal is to avoid introducing or enforcing requirements in legacy code. That way we avoid breaking existing user code without introducing a potential source of confusion.
2. Enforce the `LegacyInputIterator` requirement. If we go down this path, I would like to add a `static_assert()` at the entry point for `std::set_intersection()`, to help users identify the source of the new failure.

What do you think?

ichaer replied:
@ldionne, sorry for nagging, but could you please give feedback on this and on the PR in general?

ldionne commented on libcxx/include/__algorithm/set_intersection.h:
```diff
@@ -39,9 +44,139 @@ struct __set_intersection_result {
+        std::forward_iterator_tag,
+        std::forward_iterator_tag) {
+  std::__set_intersector<_AlgPolicy, _Compare, _InForwardIter1, _Sent1, _InForwardIter2, _Sent2, _OutIter>
+      __intersector(__first1, __last1, __first2, __last2, __result, __comp);
```
I think we should try to write this as a function instead. I believe there will be ways to reduce the duplication/complexity in that function without using a class, which IMO makes it harder to see the flow of this algorithm in its current state.

ldionne commented on libcxx/benchmarks/algorithms/set_intersection.bench.cpp:
```diff
@@ -0,0 +1,223 @@
+//===----------------------------------------------------------------------===//
```
Can you provide benchmark results in your PR description? We generally do that for optimizations.

ldionne commented on libcxx/include/__algorithm/set_intersection.h:
```diff
@@ -39,9 +44,139 @@ struct __set_intersection_result {
+  }
+};
+
+// with forward iterators we can use binary search to skip over entries
```
This comment should also explain why this satisfies the complexity guarantees of `set_intersection`. Something like:

```
When we have two forward iterators, we know we have the ability to do multiple passes over the input ranges. Therefore, we can use the one-sided lower bound algorithm to BLA BLA BLA (basically lines 89 - 106 above).
```

ldionne reviewed (changes requested):
Thanks a lot for the patch! I have some feedback but I think this should be fine to move forward.

ichaer committed:
Address feedback about qualifying abort(), added comment to clarify choice of not having a `default` case in `switch`.

ichaer committed:
Address comment about broken comment for `getVectorOfRandom()`: move the function closer to its point of usage and document what `genCacheUnfriendlyData()` is trying to do in its own comment. `getVectorOfRandom()` has imho a good name which describes all it's meant to achieve, it's `genCacheUnfriendlyData()` that needs explaining.

ichaer committed:
Merge remote-tracking branch 'llvm/main' into onesided_lower_bound

* llvm/main: (7814 commits)
  [libc++] Add some private headers to libcxx.imp (#89568)
  [Frontend][OpenMP] Add functions for checking construct type (#87258)
  AtomicExpand: Emit or with constant on RHS
  [VPlan] Ignore incoming values with constant false mask. (#89384)
  [Clang][Parser] Don't always destroy template annotations at the end of a declaration (#89494)
  [X86] getTargetShuffleMask - update to take a SDValue instead of a SDNode. NFC.
  [DAGCombiner] Fix miscompile bug in combineShiftOfShiftedLogic (#89616)
  [DAGCombiner] Pre-commit test case for miscompile bug in combineShiftOfShiftedLogic
  [Flang][OpenMP] Add restriction about subobjects to firstprivate and … (#89608)
  [SelectionDAG] Mark frame index as "aliased" at argument copy elison (#89712)
  Pre-commit reproducer for argument copy elison related bug
  [InstCombine] Fold fcmp into select (#86482)
  Make default initialization explicit
  [clang-tidy] Avoid overflow when dumping unsigned integer values (#85060)
  [ARM][AArch64] Split out processor and feature tablegen defs [NFC] (#88282)
  [TableGen] Fix ReplaceRegAction RTTI Kind
  [VectorCombine] foldShuffleOfShuffles - add missing arguments to getShuffleCost calls.
  [VPlan] Skip extending ICmp results in trunateToMinimalBitwidth.
  [LLVM][CodeGen][SVE] rev(whilelo(a,b)) -> whilehi(b,a). (#88294)
  RenameIndependentSubregs: Add missing sub-range for new IMPLICIT_DEFs (#89050)
  ...

ichaer committed:
s/_LIBCPP_NODISCARD_EXT/_LIBCPP_NODISCARD/ after merging #87094

ichaer committed:
Oops, forgot to format =/. Still working on the remaining feedback, but it would be good to be sure that we have a good baseline after this big merge from main.

ichaer committed:
Address comment about making the benchmark's `struct MoveInto` into a function -- make it a lambda, to avoid the explicit template parameter a freestanding function would require.

ichaer committed:
Address comment about using `common.h`'s `Quantities` constant in the benchmark.

ichaer committed:
Address feedback to improve assertion in _IterOps::__advance()

ichaer committed:
Rename new sentinel-based `_IterOps::advance()` to `_IterOps::__advance_to` -- no reason IMO to have a second override if `__advance_to = ranges::advance` in c++20...

ichaer committed:
Address feedback about using `iterator_traits<_Iter>::difference_type` instead of a templated `_Distance` in `_IterOps::__advance_to()`

ichaer committed:
git clang-format on the last batch of changes

ichaer committed:
Address review comments about lower_bound.h

ichaer committed:
Address review comments about set_intersection.h: unnecessary namespace qualification, insufficient comments, and direct use of iterator traits.

ichaer committed:
Address review comment about replacing `struct __set_intersector` with a function. I think I managed to preserve readability by keeping `__add_output_unless()` as a lambda.

ichaer committed:
Make `__add_output_unless()` a freestanding function, `__set_intersection_add_output_unless()`, because the lambda [tripped the "MacOS with C++03" test run](https://buildkite.com/llvm-project/libcxx-ci/builds/35055#018f173c-f155-4fbb-b6d7-a7aba01cec9e):
```
```

ichaer committed:
Address comment about using ` std::forward<_Compare>()` for consistency in `__set_intersection()` base overload.

ichaer committed:
Address review feedback: remove benchmark counters.

ichaer committed:
clang-format of the last 2 changes

ichaer commented on libcxx/include/__algorithm/set_intersection.h:
```diff
@@ -55,12 +190,41 @@ __set_intersection(
+template <class _AlgPolicy, class _Iter>
+class __set_intersection_iter_category {
+  template <class _It>
+  using __cat = typename std::_IterOps<_AlgPolicy>::template __iterator_category<_It>;
```
It worked fine =/. I must have misinterpreted compiler errors at some point... thanks for flagging that mistake, the proposal looks much cleaner now.

ichaer commented on libcxx/benchmarks/algorithms/set_intersection.bench.cpp:
```diff
@@ -0,0 +1,223 @@
+      while (state.KeepRunningBatch(BATCH_SIZE)) {
+        for (unsigned i = 0; i < BATCH_SIZE; ++i) {
+          const auto& [c1, c2] = input;
+          if (TRACK_COUNTERS) {
```
I've removed it from this PR, I'll give it some thought to see if I can come up with something like what you described... it's tricky because there's no way additional counters won't in some way impact the measurements -- a bit like the observer effect :) --, which is why I gated counter collection on user input, and disabled it by default: time measurements should ideally not be collected at the same time as counter output. It's a tricky tension between fidelity and detail.

ichaer reviewed (commented):
I think I've addressed all review comments. I know you're out this week, there's no rush, but please let me know what you think once you've had the time to take a look :)

ichaer committed:
Oops, leftover template type name!

ichaer committed:
Remove unnecessary PauseTiming()/ResumeTiming() in the benchmark data generation stage, time won't be measured before we go into the benchmark::State loops.

ichaer commented:
I've deleted my old comment about inconsistent benchmark results to avoid misleading others with my bad conjectures. The real reason benchmark results were inconsistent was the fact that we use randomly-generated data as input, and, embarrassingly, it hadn't occurred to me to seed the random number generators involved with hardcoded values before producing results for a before-and-after comparison. Given a large enough sample, and if we were looking only at the results in aggregate, we'd expect that not to make a difference, but samples were not large enough (especially for the larger runs), and I was comparing test cases independently, so comparison results were surprising at times.

What I've done to produce the comparison results shared in the PR comment was use a hardcoded value of 0 as the seed for `getRandomEngine()` in `libcxx/benchmarks/GenerateInput.h`. I'm attaching the json benchmark results here in case anyone is interested in looking at the full thing without having to run it themselves.

[onesided-be6c5c848d.json](https://github.com/llvm/llvm-project/files/15163968/onesided-be6c5c848d.json)
[main-b926f75e89.json](https://github.com/llvm/llvm-project/files/15163972/main-b926f75e89.json)

ichaer committed:
Apply suggestions from code review

@ldionne's latest inline review suggestions.

Co-authored-by: Louis Dionne <ldionne.2@gmail.com>

ichaer committed:
Move new `_IterOps<_ClassicAlgPolicy>::__advance_to()` overloads next to its pre-existing sibling and remove leftover comments.

ichaer committed:
Refactor operation counts out of `stride_counting_iterator`.

ichaer committed:
clang-format fixups to inline code suggestions

ichaer committed:
change complexity test interface, 's/testSetIntersectionAndReturnOpCounts/counted_set_intersection/', and split complexity tests into `set_intersection_complexity.pass.cpp`.

ichaer committed:
's/stride_counting_iterator/operation_counting_iterator/'

ichaer committed:
Move last complexity test to the new file, and add a matching one for `std::set_intersection()`

ichaer committed:
Add yet another complexity test, this one validating the standard guarantees for a single match over an input of 1 to 20 elements.

ichaer commented on libcxx/test/std/algorithms/alg.sorting/alg.set.operations/set.intersection/ranges_set_intersection.pass.cpp:
```diff
@@ -272,6 +278,234 @@ constexpr void runAllIteratorPermutationsTests() {
+      In2<decltype(in2.begin())>(in2.begin()), &ops.in[1].iterator_strides, &ops.in[1].iterator_displacement);
+  stride_counting_iterator e2(
+      In2<decltype(in2.end()) >(in2.end()), &ops.in[1].iterator_strides, &ops.in[1].iterator_displacement);
+
```
I went for option number 2.

ichaer commented on libcxx/test/support/test_iterators.h:
```diff
@@ -733,8 +733,24 @@ struct common_input_iterator {
+template <class It,
+          class StrideCountType        = std::iter_difference_t<It>,
+          class StrideDisplacementType = std::iter_difference_t<It>>
 class stride_counting_iterator {
```
I did it in this change, I think the size is OK, but let me know what you think.

ichaer committed:
Take into account additional comparison performed in assertion in hardening mode inside `testComplexityBasic()` as well.

ichaer committed:
Add release note. It reads a bit awkward, maybe I'll come up with something better after some more thinking...

ldionne commented on libcxx/include/__algorithm/set_intersection.h:
```diff
@@ -38,10 +43,94 @@ struct __set_intersection_result {
+    _InForwardIter1 __first1_next =
+        std::__lower_bound_onesided<_AlgPolicy>(__first1, __last1, *__first2, __comp, __proj);
+    std::swap(__first1_next, __first1);
+    std::__set_intersection_add_output_unless(__first1 != __first1_next, __first1, __first2, __result, __prev_advanced);
```
I think the algorithm would be easier to read if we inlined this function at its two call sites. Why do we need to keep track of whether we previously advanced the iterators?

ichaer replied:
Oh, maybe I should have documented that! It's a (hopefully :)) clever optimisation to preserve the number of comparisons. In `set_intersection()` we compute `a==b` by means of `!(a<b || b<a)`, right? We can avoid further comparison operator calls by checking whether we've advanced any side. If, at any time, we have two subsequent attempts to advance iterators failing, then we know elements are equal.

ichaer replied:
(Maybe simply calling it an optimisation is unfair. This is what allows us to preserve the complexity guarantees from the current standard.)

ldionne commented on libcxx/include/__algorithm/set_intersection.h:
```diff
@@ -38,10 +43,94 @@ struct __set_intersection_result {
+    std::forward_iterator_tag,
+    std::forward_iterator_tag) {
+  _LIBCPP_CONSTEXPR std::__identity __proj;
+  bool __prev_advanced = true;
```
It feels off that we start with `__prev_advanced = true`. I'm not saying the algorithm's wrong, but is it possible that another name for this variable is trying to emerge?

ichaer replied:
Maybe. Would you be happier with something like `__maybe_equals`? I'd have to invert the logic, but maybe that would help make it more readable?

ldionne replied:
I think it might help, yes!

ichaer replied:
Cool, I've renamed the variables and added a comment at the point of call which I'm hoping captures what I said here :). Let me know what you think, and whether you'd still prefer for the code inside `__set_intersection_add_output_if_equal()` to be inlined. My personal feeling about it is that the function wraps a fiddly idea, and that it's easier for me to unpack the algorithm when I give that idea a name, but I defer to you.

ldionne reviewed (commented):
This basically LGTM but I have a few small comments left.

philnik777 commented on libcxx/include/__algorithm/lower_bound.h:
```diff
@@ -46,6 +48,48 @@ __lower_bound(_Iter __first, _Sent __last, const _Type& __value, _Comp& __comp,
+
+template <class _AlgPolicy, class _ForwardIterator, class _Sent, class _Type, class _Proj, class _Comp>
+_LIBCPP_NODISCARD inline _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 _ForwardIterator
+__lower_bound(_ForwardIterator __first, _Sent __last, const _Type& __value, _Comp& __comp, _Proj& __proj) {
```
We can certainly improve the complexity here (http://eel.is/c++draft/structure#specifications-7). I do wonder whether this is also the case for the "exactly N times" specifications, but I don't think it's in the interest of anybody to force implementations to add overhead for conformance. Making the complexity worse (by more than a constant factor) would be another story. We already break the complexity requirements by a constant factor if it's not observable, e.g. in `ranges::minmax` with integers.

ldionne replied:
That's also my thinking.

@ichaer After this patch has landed, you could consider a patch like the above if you want -- I think we'd take it.

ichaer replied:
Awesome =D, thanks!

ichaer committed:
* s/__(prev_)?advanced/__$1_may_be_equal/g
* s/__set_intersection_add_output_unless/__set_intersection_add_output_if_equal/
* Add comments to explain the logic for equality comparison

ichaer committed:
Oops

ichaer committed:
Merge remote-tracking branch 'llvm/main' into onesided_lower_bound

* llvm/main: (8718 commits)
  [LLVM] Add `llvm.experimental.vector.compress` intrinsic (#92289)
  [Clang] [C23] Implement N2653: u8 strings are char8_t[] (#97208)
  [SLP][REVEC] Make SLP support revectorization (-slp-revec) and add simple test. (#98269)
  [Flang] Exclude the reference to TIME_UTC for AIX. (#99069)
  [clang][Sema] Improve `Sema::CheckCXXDefaultArguments` (#97338)
  adjust the Xtensa backend after change f270a4dd6667759d7305797a077ae09648318ac7
  [lldb][Bazel]: Second attempt to adapt for a751f653b40f2021f091a2f1ebcc2d91bc4cc89d
  Revert "[lldb][Bazel]: Adapt BUILD.bazel file for a751f653b40f2021f091a2f1ebcc2d91bc4cc89d"
  [lldb][Bazel]: Adapt BUILD.bazel file for a751f653b40f2021f091a2f1ebcc2d91bc4cc89d
  [emacs] Fix autoloading for llvm-mir-mode (#98984)
  [SLP]Improve minbitwidth analysis for trun'ed gather nodes.
  [Clang][Concepts] Avoid substituting into constraints for invalid TemplateDecls (#75697)
  [LV][NFC]Introduce isScalableVectorizationAllowed() to refactor getMaxLegalScalableVF().
  Revert "[AArch64] Remove superfluous sxtw in peephole opt (#96293)"
  [AArch64][GISel] Add test cases for folding shifts into load/store addressing modes (NFC)
  [LV] Process dead interleave pointer ops in reverse order.
  [AMDGPU] Remove SIWholeQuadMode pass early exit (#98450)
  [InstrRef][NFC] Avoid un-necessary DenseMap queries (#99048)
  [gn build] Port e94e72a0c229
  [LV][AArch64] Prefer Fixed over Scalable if cost-model is equal (Neoverse V2) (#95819)
  ...

ldionne reviewed (approved):
@ichaer I personally find the helper unhelpful in understanding the algorithm, but I'm OK with merging as-is. Will merge once CI is green.

Thanks for the patch and for all the iterations you did on this.

ichaer commented:
Thank you, @ldionne :).

alexfh commented:
Thanks for the nice optimization!

However, I'd like to point out one real-world complication it brings: it breaks some invalid usages of `std::set_intersection` in a hard-to-debug way. I've just investigated a test failure that boiled down to unsorted sequences passed to `std::set_intersection`. The only clue was that the failure started appearing after this libc++ commit. Without this hint I can imagine many more hours of poking around.

I wonder, if it would be possible to add runtime checks in one of the hardening modes to assist with finding this sort of an issue? For example, a check for the correct ordering of the bounds could be made on each step of the binary search. This should at least retain the complexity. A more comprehensive check (ordering of all elements in both input sequences) could be done in `_LIBCPP_HARDENING_MODE_EXTENSIVE` or `_LIBCPP_HARDENING_MODE_DEBUG` mode. What do you think about this?

An example of a problematic use: https://gcc.godbolt.org/z/Y5dxsxzbe.

ichaer commented:
@alexfh, I did think of that when proposing the change, but I couldn't think of a good way to help users. Your idea of adding validation to the hardening modes sounds good to me, I've got my hands full right now, but I'll open a PR for that as soon as I can.

The case you shared, with `{944, 3982, 10041, 443}` and `{10041}`, is easy to detect cheaply, but if the last entry were `3982` instead it would be trickier... I'm torn on this. What did you have in mind, a full scan over inputs to check if they are sorted, or the cheap but much weaker validation that we can't reach a value which is less than the previous one we've inspected?

alexfh commented:
I didn't think too much about tradeoffs here, but it seems like a full linear scan in `_LIBCPP_HARDENING_MODE_DEBUG` should be totally uncontroversial.

As for a cheaper check for `_LIBCPP_HARDENING_MODE_EXTENSIVE` or even `_LIBCPP_HARDENING_MODE_FAST`, a log(n) sampling while doing the binary search could be helpful and cheap enough, but I'd actually expect some measurements there, especially in case of `_LIBCPP_HARDENING_MODE_FAST`.

philnik777 commented:
@alexfh Can this result in any security-concerning bugs like out-of-bounds reads? If not I'd just go for a `_LIBCPP_ASSERT_WHATEVER(std::is_sorted(__first, __last, __comp), "Passed unsored range to set_intersection")` that only runs in debug mode and call it a day. Asking for people to run the debug mode when debugging stuff seems reasonable to me.

ichaer commented:
No out-of-bounds reads, we still respect the end sentinel. What happens is that we now skip over potential matches, and if the sequence is unsorted we may not go back to see what we skipped over.

ichaer commented:
@alexfh, @philnik777: I've opened https://github.com/llvm/llvm-project/pull/101508 with what we discussed.
